rfd/0057-automatic-user-provisioning.md

```diff
@@ -0,0 +1,252 @@
+version: v5
+metadata:
+  name: example
+spec:
+  options:
+    # Controls whether this role supports auto provisioning of users.
+    create_host_user: true
+  allow:
+    # New field listing Linux groups to assign a provisioned user to.
+    # Should support user and identity provider traits like other fields (e.g. "logins")
+    host_groups: [ubuntu, "{{internal.groups}}", "{{external.xxx}}"]
+    # host_sudoers is a list of entries to be included in a users sudoers file
+    host_sudoers: ["{{internal.logins}} ALL=(ALL) ALL", ...]
+```
+
+An individual `ssh_service` can be configured disable auto user
+creation with the below config:
+
+```yaml
+ssh_service:
+    # when disabled, takes precedence over the role setting
+    disable_create_host_user: true
+```
+
+### User creation
+
+In order to create users `useradd` will be executed from teleport
+after a user has tried to access a Teleport SSH node.
+
+#### User Groups
+
+When a user is created they will be added to the specified groups from
+the `host_groups` field in the role. In addition the user will be
+added to a special `teleport-system` group which can be used to
+indicate that the user was created by teleport and that its safe for
+it to be deleted. The groups will be created via `groupadd` at startup
+if they do not already exist and users will be added to groups via
+`usermod -aG <list of groups> <username>`
+
+#### Valid user/group names
+
+The set of valid names that are valid on Linux varies between distros
+and are generally more restrictive than the allowed usernames in
+Teleport. This will require that names containing invalid characters
+have those characters removed/replaced. Information on the valid
+characters between Linux distros is available [here](https://systemd.io/USER_NAMES/).
+The common core of valid characters is `^[a-z][a-z0-9-]{0,30}$`.
+
+#### Adding and removing users from sudoers
+
+Each user with entries in `host_sudoers` will have a file created in
+`/etc/sudoers.d`, with one entry per line.
+
+If a user is in multiple rules that specify `host_sudoers` they will
+be all be concatenated together.
+
+##### sudoers file syntax validation
+
+If a system has `visudo` present, validation could be performed by
+executing `visudo -c -f path/to/sudoersfile`, where if it fails to
+validate, the user fails to have the shell start and the error is
+reported.
+
+##### sudoers security considerations
+
+In order to stop users from being able to edit the sudoers file a
+command allow list must be used, as or equivalent to below:
+
+```
+${USERNAME} ALL = (${USER TO RUN AS}) NOPASSWD: /bin/cmd1 args, /bin/cmd2 args
+```
+
+Should a user be given `root` access to all commands, they will be
+able to modify any file, including sudoers files.
+
+
+### User and group deletion
+
+After all of a users sessions are logged out the created user and any
+`sudoers` files that were created for that user will be deleted if
+that user is also a member of the `teleport-system` group.
```

Comment:
As it stands I expected UIDs to be reused, but perhaps it could make sense to have user deletion set the shell to /bin/nologin or something, this way UIDs wont clash